### PR TITLE
Chat scrolling improvements

### PR DESCRIPTION
**Pull request by cwalther**
These two changes ([41765b9](https://foundry.openuru.org/gitblit/commit/?r=MOULSCRIPT-ou.git&h=41765b9f8c9feff5c1e2d135580f7a2c394a3c76) and [28f668f](https://foundry.openuru.org/gitblit/commit/?r=MOULSCRIPT-ou.git&h=28f668f6c16045a0863158989ffd0d0383202847)) have been in MOULa for a long time, but apparently I forgot to submit them back here after I packed them onto Lyrositor’s [OU equivalent](https://bitbucket.org/OpenUru_org/moulscript-ou/pull-requests/19) of #40, as I noticed last weekend on Gehn.

Presently, when KI chat text field is full, i.e. incoming lines push out old ones at the top, and you are scrolled up to somewhere in the middle of the field, an incoming line still shifts the text by one line, instead of keeping the scroll position locked to the text as intended (and as it works when the field is not full). The first commit fixes that.

This however makes incoming chat unnoticeable when scrolled up. To compensate for that, the second commit makes the down scroll arrow blink when chat comes in.

To illustrate the usability issue caused by the current behavior: At the Veelay Tsahvahn party on Gehn it happened to me once that I answered a question that had actually been posted several minutes before (and had long been answered by others), because it was just then pushed into my visible chat range, and I hadn’t noticed that I was scrolled up.

### CODE REVIEW

File: Python/ki/__init__.py [hunk 1]
```diff
@@ -1123,6 +1123,23 @@ def OnTimer(self, ID):
...
+            if self.chatMgr.incomingChatFlashState > 0:
+                btn = ptGUIControlButton(mKIdialog.getControlFromTag(kGUI.miniChatScrollDown))
+                if self.chatMgr.incomingChatFlashState & 1:
+                    btn.hide()
+                else:
+                    btn.show()
+                self.chatMgr.incomingChatFlashState -= 1
+                PtAtTimeCallback(self.key, 0.15, kTimers.IncomingChatFlash)
+            else:
+                mKIdialog.refreshAllControls()
```
Comment (Hoikas) [inline Python/ki/__init__.py]: Why does the entire KI need to be redrawn? This is an expensive operation according to my profiling (corresponding to several framerate stutters) from several years ago, and I find myself adverse to introducing unnecessary refreshes.

Comment (cwalther) [reply Python/ki/__init__.py]: Good point. I don’t remember. It probably doesn’t – I guess the only thing that needs to be redrawn in the proper shown or hidden state is the scroll arrow. I’ll take a look.

**Comment by cwalther** [comment]
On closer examination, all that needs to be refreshed is pfGUIUpDownPairMod "ChatScrollUpDown". The trouble is, I can’t easily/efficiently/reliably get hold of that, because it shares its tag ID 51 with pfGUIButtonMod "miniKIChatScollUp". So, for best performance, this new version needs a change to the PRPs:
- In *GUI_District_KIMini.prp* at offset 0x143b3b (1325883), change 0x33 to 0x37.
- In *GUI_District_KIMicro.prp* at offset 0x1e17 (7703), change 0x33 to 0x37.

Any better ideas?

**Comment by Hoikas** [comment]
I just tested that `ptGUIControlButton(mKIdialog.getControlFromTag(kGUI.miniChatScrollDown)).show()` will successfully restore the arrow state in the else clause.

**Comment by Hoikas** [comment]
In light of a discussion on IRC about the potential for the user to scroll down while the arrows are flashing, and considering how trivial this case is, it's probably best to refresh the entire KI dialog. Even though we are versioning the assets, it's still difficult to redistribute them outside of a "playing" situation. I'm afraid that more likely than not, the proposed workaround for non-updated assets would become a permanent fixture in the code.

**Comment by cwalther** [comment]
Here’s the previous version back.

**Pull request merged**